zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,642 @@
+// Copyright 2021-2022 Zeitgeist PM LLC.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#![doc = include_str!("../README.md")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarks;
+mod global_disputes_pallet_api;
+mod mock;
+mod tests;
+pub mod types;
+pub mod weights;
+
+pub use global_disputes_pallet_api::GlobalDisputesPalletApi;
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{types::*, weights::WeightInfoZeitgeist, GlobalDisputesPalletApi};
+    use core::marker::PhantomData;
+    use frame_support::{
+        ensure, log,
+        pallet_prelude::{
+            DispatchResultWithPostInfo, OptionQuery, StorageDoubleMap, StorageMap, ValueQuery,
+        },
+        sp_runtime::traits::StaticLookup,
+        traits::{
+            Currency, ExistenceRequirement, Get, IsType, LockIdentifier, LockableCurrency,
+            WithdrawReasons,
+        },
+        Blake2_128Concat, BoundedVec, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use sp_runtime::{
+        traits::{AccountIdConversion, CheckedDiv, Saturating, Zero},
+        DispatchResult, SaturatedConversion,
+    };
+    use sp_std::{vec, vec::Vec};
+    use zeitgeist_primitives::types::OutcomeReport;
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The currency implementation used to lock tokens for voting.
+        type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// The vote lock identifier.
+        #[pallet::constant]
+        type GlobalDisputeLockId: Get<LockIdentifier>;
+
+        /// The pallet identifier.
+        #[pallet::constant]
+        type GlobalDisputesPalletId: Get<PalletId>;
+
+        /// To reference the market id type.
+        type MarketCommons: MarketCommonsPalletApi<
+            AccountId = Self::AccountId,
+            BlockNumber = Self::BlockNumber,
+        >;
+
+        /// The maximum numbers of distinct markets
+        /// on which one account can simultaneously vote on outcomes.
+        #[pallet::constant]
+        type MaxGlobalDisputeVotes: Get<u32>;
```

Comment:
It's to save storage for the number of locks. Otherwise users can just keep voting on different global disputes and never unlock. When the user unlocks, the user has again `MaxGlobalDisputeVotes` number of votes.

https://github.com/zeitgeistpm/zeitgeist/blob/79b3f31adbdec8a96ed13b916972de8ca616cedb/zrml/global-disputes/src/lib.rs#L458

Reply:
1338b1f79e882affd509180fe9288af97b3f6b00